packages/matter-node-ble.js/src/ble/BlenoBleServer.ts

```diff
@@ -0,0 +1,352 @@
+
+            // callback to disconnect the BLE connection
+            async () => this.close(),
+
+            // callback to forward decoded and de-assembled Matter messages to ExchangeManager
+            async (data: ByteArray) => {
+                if (this.onMatterMessageListener === undefined) {
+                    throw new Error(`No listener registered for Matter messages`);
+                }
+                this.onMatterMessageListener(this, data);
+            }
+        );
+        this.latestHandshakePayload = undefined; // BTP Session initialized, handshake payload not needed anymore
+    }
+
+    handleC2Indicate() {
+        if (this.writeConformationResolver !== undefined) {
+            this.writeConformationResolver();
+            this.writeConformationResolver = undefined;
+        } else {
+            logger.warn(`Received C2 indication but no former write expected a confirmation`);
+        }
+    }
+
+    /**
+     * Process a Read request on characteristic C3 from the Matter service.
+     * The relevant data needs optionally to be set before advertising, else empty data are used.
+     *
+     * @param offset
+     */
+    handleC3ReadRequest(offset: number) {
+        if (offset > this.additionalAdvertisingData.length) {
+            throw new Error(`Offset ${offset} is larger than data ${this.additionalAdvertisingData.length}`);
+        } else {
+            return this.additionalAdvertisingData.subarray(offset);
+        }
+    }
+
+    advertise(advertiseData: ByteArray, additionalAdvertisementData?: ByteArray) {
+        this.advertisingData = Buffer.from(advertiseData.buffer);
+
+        if (additionalAdvertisementData) {
+            this.additionalAdvertisingData = Buffer.from(additionalAdvertisementData.buffer);
+        } else {
+            this.additionalAdvertisingData = Buffer.alloc(0);
+        }
+
+        if (this.isAdvertising) {
+            Bleno.stopAdvertising();
+            this.isAdvertising = false;
+        }
+
+        if (this.state === 'poweredOn') {
+            Bleno.startAdvertisingWithEIRData(this.advertisingData);
+            this.isAdvertising = true;
+        } else {
+            logger.debug(`State is ${this.state}, advertise when powered on`);
+        }
+    }
+
+    stopAdvertising() {
+        if (this.isAdvertising) {
+            Bleno.stopAdvertising();
+            this.isAdvertising = false;
+        }
+    }
+
+    setMatterMessageListener(listener: (socket: Channel<ByteArray>, data: ByteArray) => void) {
+        if (this.onMatterMessageListener !== undefined) {
+            throw new Error(`onData listener already set`);
+        }
+        this.onMatterMessageListener = listener;
+    }
+
+    btpHandshakeTimeoutTriggered() {
+        Bleno.disconnect();
+        logger.error("Timeout for handshake subscribe request on C2 reached, disconnecting.");
+    }
+
+    async close() {
+        this.btpHandshakeTimeout.stop();
+        Bleno.disconnect();
```

Comment:
Should we create a promise, add a Bleno "onDisconnect" event handler, then await the promise here?

Reply:
good idea, added